lib/handlers/authorize-handler.js

```diff
@@ -0,0 +1,310 @@
+
+/**
+ * Module dependencies.
+ */
+
+var _ = require('lodash');
+var AccessDeniedError = require('../errors/access-denied-error');
+var AuthenticateHandler = require('../handlers/authenticate-handler');
+var InvalidArgumentError = require('../errors/invalid-argument-error');
+var InvalidClientError = require('../errors/invalid-client-error');
+var InvalidRequestError = require('../errors/invalid-request-error');
+var InvalidScopeError = require('../errors/invalid-scope-error');
+var OAuthError = require('../errors/oauth-error');
+var Promise = require('bluebird');
+var Request = require('../request');
+var Response = require('../response');
+var ServerError = require('../errors/server-error');
+var UnauthorizedClientError = require('../errors/unauthorized-client-error');
+var is = require('../validator/is');
+var tokenUtil = require('../utils/token-util');
+var url = require('url');
+
+/**
+ * Response types.
+ */
+
+var responseTypes = {
+  code: require('../response-types/code-response-type'),
+  token: require('../response-types/token-response-type')
+};
+
+/**
+ * Constructor.
+ */
+
+function AuthorizeHandler(options) {
+  options = options || {};
+
+  if (!options.authorizationCodeLifetime) {
+    throw new InvalidArgumentError('Missing parameter: `authorizationCodeLifetime`');
+  }
+
+  if (!options.model) {
+    throw new InvalidArgumentError('Missing parameter: `model`');
+  }
+
+  if (!options.model.getClient) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `getClient()`');
+  }
+
+  if (!options.model.saveAuthorizationCode) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `saveAuthorizationCode()`');
+  }
+
+  this.authorizationCodeLifetime = options.authorizationCodeLifetime;
+  this.authenticateHandler = new AuthenticateHandler(options);
+  this.model = options.model;
+}
+
+/**
+ * Authorize Handler.
+ */
+
+AuthorizeHandler.prototype.handle = function(request, response) {
+  if (!(request instanceof Request)) {
+    throw new InvalidArgumentError('Invalid argument: `request` must be an instance of Request');
+  }
+
+  if (!(response instanceof Response)) {
+    throw new InvalidArgumentError('Invalid argument: `response` must be an instance of Response');
+  }
+
+  if (false === request.body.allowed) {
+    return Promise.reject(new AccessDeniedError('Access denied: user denied access to application'));
+  }
+
+  var fns = [
+    this.generateAuthorizationCode(),
+    this.getAuthorizationCodeLifetime(),
+    this.getClient(request),
+    this.getUser(request, response)
+  ];
+
+  return Promise.all(fns)
```

Comment:
Hmmm, I see what you mean. Yes, I can make that change but the same can also be said of the other methods (`generateAuthorizationCode()`, `getAuthorizationCodeLifetime()` and `getUser()`).

If any of them fail, the entire request will certainly fail. Since each of these methods returns a promise, I'd have to nest all of these in a 4-level deep chain which was what I was trying to avoid.

However, I understand the overhead argument and will look into making the change. This will all become a bit easier to manage once/if we start using ES6 _async/await_